File: app/src/main/java/com/dimatest/movieapp/common/BaseUseCaseSingle.java
package com.dimatest.movieapp.common;

import androidx.core.util.Preconditions;

import io.reactivex.Scheduler;
import io.reactivex.Single;
import io.reactivex.observers.DisposableSingleObserver;

public abstract class BaseUseCaseSingle<T, Params> extends BaseUseCase {

    public BaseUseCaseSingle(Scheduler mainThread, Scheduler ioThread) {
        super(mainThread, ioThread);
    }

    abstract protected Single<T> buildUseCaseObservable(Params params);

    public void execute(Params params, DisposableSingleObserver<T> observer) {
        Preconditions.checkNotNull(observer);
        final Single<T> observable = this.buildUseCaseObservable(params)
                .subscribeOn(ioThread)
                .observeOn(mainThread)
                .doOnSubscribe(__ -> isLoading.postValue(true))
                .doOnTerminate(() -> isLoading.postValue(false))
                .doOnDispose(() -> isLoading.postValue(false));
        addDisposable(observable.subscribeWith(observer));
    }
}

File: app/src/main/java/com/dimatest/movieapp/common/BaseUseCaseSingle.kt
package com.dimatest.movieapp.common

import android.util.Log
import io.reactivex.Scheduler
import io.reactivex.Single
import io.reactivex.observers.DisposableSingleObserver

abstract class BaseUseCaseSingle<T, Params>(mainThread: Scheduler, ioThread: Scheduler) : BaseUseCase(mainThread, ioThread) {
    protected abstract fun buildUseCaseObservable(params: Params): Single<T>

    private fun execute(params: Params, observer: DisposableSingleObserver<T>) {
        val observable = buildUseCaseObservable(params)
                .subscribeOn(ioThread)
                .observeOn(mainThread)
                .doOnSubscribe { isLoading.postValue(true) }
                .doOnTerminate { isLoading.postValue(false) }
                .doOnDispose { isLoading.postValue(false) }
                .onErrorResumeNext { Single.error(it) }
        addDisposable(observable.subscribeWith(observer))
    }

    fun execute(
            params: Params,
            onSuccess: ((t: T) -> Unit)? = null,
            onError: ((Throwable) -> Unit)? = null
    ) {
        execute(params, object : DisposableSingleObserver<T>() {
            override fun onSuccess(t: T) {
                onSuccess?.invoke(t)
            }

            override fun onError(e: Throwable) {
                Log.d(this.javaClass.name, e.toString())
                onError?.invoke(e)
            }
        })
    }
}